core/vdbe.rs

```diff
@@ -192,9 +202,15 @@ impl ProgramState {
     }
 }
 
+pub enum ProgramType {
+    Default,
+    PragmaChange(String, i64),
+}
```

Comment:
Let's not add a special-purpose program for pragmas. For example, `journal_mode` seems to be totally handled by specific opcodes like this:

```
sqlite> EXPLAIN pragma journal_mode = wal;
addr  opcode         p1    p2    p3    p4             p5  comment
----  -------------  ----  ----  ----  -------------  --  -------------
0     Init           0     5     0                    0   Start at 5
1     Expire         1     1     0                    0
2     JournalMode    0     1     5                    0
3     ResultRow      1     1     0                    0   output=r[1]
4     Halt           0     0     0                    0
5     Goto           0     1     0                    0
```

In fact, even for `cache_size` pragma, normal VDBE programs are generated:

```
sqlite> EXPLAIN pragma cache_size = 100;
addr  opcode         p1    p2    p3    p4             p5  comment
----  -------------  ----  ----  ----  -------------  --  -------------
0     Init           0     3     0                    0   Start at 3
1     Expire         1     1     0                    0
2     Halt           0     0     0                    0
3     Goto           0     1     0                    0
```

```
sqlite> EXPLAIN pragma cache_size;
addr  opcode         p1    p2    p3    p4             p5  comment
----  -------------  ----  ----  ----  -------------  --  -------------
0     Init           0     5     0                    0   Start at 5
1     Expire         1     1     0                    0
2     Int64          0     1     0     100            0   r[1]=100
3     ResultRow      1     1     0                    0   output=r[1]
4     Halt           0     0     0                    0
5     Goto           0     1     0                    0
```

It's just that for this specific pragma, the changes are applied _before_ program execution:

```
  case PragTyp_CACHE_SIZE: {
    assert( sqlite3SchemaMutexHeld(db, iDb, 0) );
    if( !zRight ){
      returnSingleInt(v, pDb->pSchema->cache_size);
    }else{
      int size = sqlite3Atoi(zRight);
      pDb->pSchema->cache_size = size;
      sqlite3BtreeSetCacheSize(pDb->pBt, pDb->pSchema->cache_size);
    }
    break;
  }
```

I would like to see us just follow SQLite here.

Reply:
Sadly, persistence of `cache_size` is wrong. `default_cache_size` is different from `cache_size` and you can only persist the default:
```
sqlite> explain pragma default_cache_size;
addr  opcode         p1    p2    p3    p4             p5  comment      
----  -------------  ----  ----  ----  -------------  --  -------------
0     Init           0     12    0                    0   Start at 12
1     Expire         1     1     0                    0   
2     Transaction    0     0     0                    0   
3     ReadCookie     0     1     3                    0   
4     IfPos          1     10    0                    0   if r[1]>0 then r[1]-=0, goto 10
5     Integer        0     2     0                    0   r[2]=0
6     Subtract       1     2     1                    0   r[1]=r[2]-r[1]
7     IfPos          1     10    0                    0   if r[1]>0 then r[1]-=0, goto 10
8     Integer        -2000  1     0                    0   r[1]=-2000
9     Noop           0     0     0                    0   
10    ResultRow      1     1     0                    0   output=r[1]
11    Halt           0     0     0                    0   
12    Goto           0     1     0                    0   
```
I think I might borrrow write path used here for another pr. Anyways, I've updated this part of the code